src/analysis/complex/cauchy_integral.lean

```diff
@@ -0,0 +1,363 @@
+  set t : set (ℝ × ℝ) := e ⁻¹' s,
+  rw [interval_swap z.im] at Hc Hd Hi,
+  have hR : e ⁻¹' (re ⁻¹' [z.re, w.re] ∩ im ⁻¹' [w.im, z.im]) = R := rfl,
+  have htc : ∀ p ∈ t, continuous_within_at F R p,
+    from λ p hp, (Hc (e p) hp).comp e.continuous_within_at hR.ge,
+  have htd : ∀ p ∈ R \ t, has_fderiv_within_at F (F' p) R p,
+    from λ p hp, (Hd (e p) hp).comp p e.has_fderiv_within_at hR.ge,
+  simp_rw [← interval_integral.integral_smul, interval_integral.integral_symm w.im z.im,
+    ← interval_integral.integral_neg, ← hF'],
+  refine (integral2_divergence_prod_of_has_fderiv_within_at_off_countable
+      (λ p, -(I • F p)) F (λ p, - (I • F' p)) F' z.re w.im w.re z.im t (hs.preimage e.injective)
+      (λ p hp, (continuous_within_at_const.smul (htc p hp)).neg) htc
+      (λ p hp, ((htd p hp).const_smul I).neg) htd _).symm,
+  rw [← volume_preserving_equiv_real_prod.symm.integrable_on_comp_preimage
+    (measurable_equiv.measurable_embedding _)] at Hi,
+  simpa only [hF'] using Hi.neg
+end
+
+/-- **Cauchy theorem**: the integral of a complex differentiable function over the boundary of a
+rectangle equals zero.
+
+Moreover, the same is true if `f` is only differentiable at points outside of a countable set `s`
+and is continuous at the points of this set. -/
+lemma integral_boundary_rect_eq_zero_of_differentiable_on_off_countable (f : ℂ → E)
+  (z w : ℂ) (s : set ℂ) (hs : countable s)
+  (Hc : ∀ x ∈ s, continuous_within_at f (re ⁻¹' [z.re, w.re] ∩ im ⁻¹' [z.im, w.im]) x)
+  (Hd : ∀ x ∈ (re ⁻¹' [z.re, w.re] ∩ im ⁻¹' [z.im, w.im]) \ s, differentiable_within_at ℂ f
+    (re ⁻¹' [z.re, w.re] ∩ im ⁻¹' [z.im, w.im]) x) :
+  (∫ x : ℝ in z.re..w.re, f (x + z.im * I)) - (∫ x : ℝ in z.re..w.re, f (x + w.im * I)) +
+    (I • ∫ y : ℝ in z.im..w.im, f (re w + y * I)) -
+      I • ∫ y : ℝ in z.im..w.im, f (re z + y * I) = 0 :=
+by refine (integral_boundary_rect_of_has_fderiv_within_at_real_off_countable f
+  (λ z, (fderiv_within ℂ f _ z).restrict_scalars ℝ) z w s hs Hc
+  (λ x hx, (Hd x hx).has_fderiv_within_at.restrict_scalars ℝ) _).trans _;
+    simp [← continuous_linear_map.map_smul]
+
+/-- If `f : ℂ → E` is complex differentiable on the closed annulus `r ≤ ∥z - c∥ ≤ R`, `0 < r ≤ R`,
+then the integrals of `f z / (z - c)` (formally, `(z - c)⁻¹ • f z`) over the circles
+`∥z - c∥ = r` and `∥z - c∥ = R` are equal to each other.
+
+Moreover, the same is true if `f` is differentiable at points of the annulus outside of a countable
+set `s` and is continuous at points of this set.  -/
+lemma circle_integral_sub_center_inv_smul_eq_of_differentiable_on_annulus_off_countable
+  {c : ℂ} {r R : ℝ} (h0 : 0 < r) (hle : r ≤ R) {f : ℂ → E} {s : set ℂ} (hs : countable s)
+  (hc : ∀ z ∈ s, continuous_within_at f (closed_ball c R \ ball c r) z)
+  (hd : ∀ z ∈ (closed_ball c R \ ball c r) \ s,
+    differentiable_within_at ℂ f (closed_ball c R \ ball c r) z) :
+  ∮ z in C(c, R), (z - c)⁻¹ • f z = ∮ z in C(c, r), (z - c)⁻¹ • f z :=
```

Comment:
I am not completely satisfied with the assumptions of this lemma (and you also mention this below). The assumption `hc` and `hd` require continuity within the annulus at all points of the annulus, and differentiability outside of `s`. So, you could rephrase `hc` as `continuous_on f (closed_ball c R \ ball c r)`, not weakening the result. But then it would be natural in `hd`to assume differentiability only inside the open annulus, not on the boundary. In fact, this stronger version follows from the one you have written, by applying it to `r' > r` and `R' < R`, and then letting them tend to `r` and `R`.

I think it matters because once you have strengthened this lemma, then you can strengthen similarly all the next ones in the file, at virtually no cost.

Reply:
There are two possible directions of generalization:
* the one you've outlined (probably, this should be done starting with the divergence theorem);
* require that `f z = o((z - c)⁻¹)` whenever `c ∈ s`; this requires some refactoring but I don't want to close this path.

One possible way to deal with this is to define (I skip all typeclass assumptions)
```lean
def is_divergence_theorem_fun (k : Type*) {E F} (f : E → F) (s : set E) : Prop :=
continuous_on f s ∧ ∃ t : set E, countable t ∧ ∀ x ∈ interior s \ t, differentiable_within_at k f s x
```
then replace it with
```lean
def is_divergence_theorem_fun (k : Type*) {E F} (f : E → F) (s : set E) : Prop :=
continuous_on f s ∧ ∃ t : set E, countable t ∧ ∀ x ∈ interior s \ t, differentiable_within_at k f s x ∨ ...
```
once we're ready. The only public "output" of `h : is_divergence_theorem_fun k f s` should be the divergence theorem (and things like `h.image`/`h.preimage`/`h.restrict_scalars`). There can be a few public constructors.

Reply:
I agree that there are several possible nice variations. What I really want is that, from the start, the Cauchy formula should be available with the standard assumptions (differentiable inside the open disk, continuous on the closed disk), because all the subsequent theorems will depend on this context.